From f35a0181d42daf5b65011db1842305c74bcf2685 Mon Sep 17 00:00:00 2001
From: Ben Menchaca <ben.menchaca@codeaurora.org>
Date: Fri, 7 Jun 2013 10:23:09 -0500
Subject: [ag71xx] switch to pointer sentinel rings

Change to a ring structure that reduces the number of instructions per
entry, and reduces the amount of state needed per entry in preparation
for future alignment opportunities.

Signed-off-by: Ben Menchaca <ben.menchaca@codeaurora.org>
---
 drivers/net/ethernet/atheros/ag71xx/ag71xx.h      |  26 ++++--
 drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c | 102 +++++++++++-----------
 2 files changed, 68 insertions(+), 60 deletions(-)

Index: linux-3.3.8/drivers/net/ethernet/atheros/ag71xx/ag71xx.h
===================================================================
--- linux-3.3.8.orig/drivers/net/ethernet/atheros/ag71xx/ag71xx.h	2014-06-13 14:17:39.795658701 -0500
+++ linux-3.3.8/drivers/net/ethernet/atheros/ag71xx/ag71xx.h	2014-06-13 14:35:50.405049841 -0500
@@ -92,20 +92,30 @@
 struct ag71xx_buf {
 	struct sk_buff		*skb;
 	struct ag71xx_desc	*desc;
-	dma_addr_t		dma_addr;
-	unsigned long		timestamp;
+	struct ag71xx_buf	*next;
+	union {
+		dma_addr_t	dma_addr;
+		unsigned long	timestamp;
+	};
 };
 
 struct ag71xx_ring {
+	/*
+	 * "Hot" fields in the data path.
+	 */
+	struct ag71xx_buf	*curr;
+	struct ag71xx_buf	*dirty;
+	uint16_t		size;
+	uint16_t		used;
+
+	/*
+	 * "Cold" fields - not used in the data path.
+	 */
 	struct ag71xx_buf	*buf;
+	uint16_t		mask;
+	uint16_t		desc_size;
 	u8			*descs_cpu;
 	dma_addr_t		descs_dma;
-	unsigned int		desc_size;
-	unsigned int		curr;
-	unsigned int		dirty;
-	unsigned int		size;
-	unsigned int		mask;
-	unsigned int		used;
 };
 
 struct ag71xx_mdio {
Index: linux-3.3.8/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c
===================================================================
--- linux-3.3.8.orig/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c	2014-06-13 14:17:39.805658547 -0500
+++ linux-3.3.8/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c	2014-06-13 14:29:19.785658441 -0500
@@ -135,8 +135,7 @@
 	struct net_device *dev = ag->dev;
 	unsigned int bytes_compl = 0;
 	unsigned int pkts_compl = 0;
-	unsigned int dirty = ring->dirty;
-	unsigned int mask = ring->mask;
+	struct ag71xx_buf *dirty = ring->dirty;
 	unsigned int used = ring->used;
 
 	if (!ring->buf) {
@@ -144,8 +143,7 @@
 	}
 
 	while (used) {
-		struct ag71xx_buf *buf = &ring->buf[dirty];
-		struct ag71xx_desc *desc = buf->desc;
+		struct ag71xx_desc *desc = dirty->desc;
 		struct sk_buff *skb;
 
 		/*
@@ -157,16 +155,14 @@
 			dev->stats.tx_errors++;
 		}
 
-		skb = buf->skb;
-		buf->skb = NULL;
+		skb = dirty->skb;
+		dirty->skb = NULL;
+		dirty = dirty->next;
 
 		bytes_compl += skb->len;
 		pkts_compl++;
 		dev_kfree_skb(skb);
 
-		dirty++;
-		dirty &= mask;
-
 		used--;
 	}
 
@@ -192,10 +188,11 @@
 
 		desc->ctrl = DESC_EMPTY;
 		buf->skb = NULL;
+		buf->next = &ring->buf[(i + 1) & mask];
 	}
 
-	ring->curr = 0;
-	ring->dirty = 0;
+	ring->curr = ring->buf;
+	ring->dirty = ring->buf;
 	ring->used = 0;
 	netdev_reset_queue(ag->dev);
 }
@@ -249,6 +246,7 @@
 		skb_reserve(skb, rx_buf_offset);
 
 		buf->skb = skb;
+		buf->next = &ring->buf[(i + 1) & mask];
 		buf->dma_addr = dma_map_single(&dev->dev, skb->data,
 					       rx_buf_size, DMA_FROM_DEVICE);
 
@@ -256,8 +254,8 @@
 		desc->ctrl = DESC_EMPTY;
 	}
 
-	ring->curr = 0;
-	ring->dirty = 0;
+	ring->curr = ring->buf;
+	ring->dirty = ring->buf;
 	ring->used = size;
 
 	return 0;
@@ -657,12 +655,10 @@
 {
 	struct ag71xx *ag = netdev_priv(dev);
 	struct ag71xx_ring *ring = &ag->tx_ring;
-	unsigned int curr = ring->curr;
-	unsigned int mask = ring->mask;
+	struct ag71xx_buf *curr = ring->curr;
+	struct ag71xx_desc *desc = curr->desc;
 	unsigned int used = ring->used;
 	unsigned int size = ring->size;
-	struct ag71xx_buf *buf = &ring->buf[curr];
-	struct ag71xx_desc *desc = buf->desc;
 	unsigned int len;
 	dma_addr_t dma_addr;
 
@@ -689,15 +685,14 @@
 	dma_addr = dma_map_single(&dev->dev, skb->data, len, DMA_TO_DEVICE);
 
 	netdev_sent_queue(dev, len);
-	buf->skb = skb;
-	buf->timestamp = jiffies;
+	curr->skb = skb;
+	curr->timestamp = jiffies;
 
 	/* setup descriptor fields */
 	desc->data = (u32)dma_addr;
 	desc->ctrl = len & DESC_PKTLEN_M;
 
-	curr++;
-	curr &= mask;
+	curr = curr->next;
 	ring->curr = curr;
 
 	used++;
@@ -829,20 +824,18 @@
 	struct ag71xx_platform_data *pdata = ag71xx_get_pdata(ag);
 	unsigned int sent = 0;
 	unsigned int bytes_compl = 0;
-	unsigned int dirty = ring->dirty;
-	unsigned int mask = ring->mask;
+	struct ag71xx_buf *dirty = ring->dirty;
+	struct ag71xx_desc *desc = dirty->desc;
 	unsigned int used = ring->used;
 
 	DBG("%s: processing TX ring\n", dev->name);
 
 	while (used) {
-		struct ag71xx_buf *buf = &ring->buf[dirty];
-		struct ag71xx_desc *desc = buf->desc;
 		struct sk_buff *skb;
 
 		if (unlikely(!(desc->ctrl & DESC_EMPTY))) {
 			if (unlikely(pdata->is_ar7240)) {
-				if (unlikely(ag71xx_check_dma_stuck(ag, buf->timestamp))) {
+				if (unlikely(ag71xx_check_dma_stuck(ag, dirty->timestamp))) {
 					schedule_work(&ag->restart_work);
 				}
 			}
@@ -851,16 +844,19 @@
 
 		ag71xx_wr(ag, AG71XX_REG_TX_STATUS, TX_STATUS_PS);
 
-		skb = buf->skb;
-		buf->skb = NULL;
+		skb = dirty->skb;
+		dirty->skb = NULL;
+
+		/* Move forward to the next descriptor */
+		dirty = dirty->next;
+		desc = dirty->desc;
+
+		/* Update stats and free the skb */
 		bytes_compl += skb->len;
 		sent++;
 
 		dev_kfree_skb(skb);
 
-		dirty++;
-		dirty &= mask;
-
 		used--;
 	}
 
@@ -892,13 +888,13 @@
 
 static int ag71xx_rx_packets(struct ag71xx *ag, struct net_device *dev, int limit)
 {
-	struct ag71xx_ring *ring = &ag->rx_ring;
 	unsigned int rx_buf_size = ag->rx_buf_size;
 	unsigned int rx_buf_offset = ag->rx_buf_offset;
-	unsigned int curr = ring->curr;
-	unsigned int mask = ring->mask;
+	struct ag71xx_ring *ring = &ag->rx_ring;
+	struct ag71xx_buf *curr = ring->curr;
+	struct ag71xx_buf *dirty = ring->dirty;
+	struct ag71xx_desc *desc = curr->desc;
 	unsigned int used = ring->used;
-	unsigned int dirty = ring->dirty;
 	unsigned int size = ring->size;
 	int done = 0;
 
@@ -916,8 +912,6 @@
 	 * Process newly received packets.
 	 */
 	while (done < limit) {
-		struct ag71xx_buf *buf = &ring->buf[curr];
-		struct ag71xx_desc *desc = buf->desc;
 		u32 desc_ctrl;
 		struct sk_buff *skb;
 		int pktlen;
@@ -935,15 +929,21 @@
 		pktlen = desc_ctrl & DESC_PKTLEN_M;
 		pktlen -= ETH_FCS_LEN;
 
-		dma_unmap_single(&dev->dev, buf->dma_addr,
-				 rx_buf_size, DMA_FROM_DEVICE);
+		dma_unmap_single(&dev->dev, curr->dma_addr, pktlen,
+				 DMA_FROM_DEVICE);
 
 		dev->last_rx = jiffies;
 		dev->stats.rx_packets++;
 		dev->stats.rx_bytes += pktlen;
 
-		skb = buf->skb;
-		buf->skb = NULL;
+		skb = curr->skb;
+		curr->skb = NULL;
+
+		/*
+		 * Move forward to the next descriptor.
+		 */
+		curr = curr->next;
+		desc = curr->desc;
 
 		/*
 		 * Set up the offset and length of the skb.
@@ -969,9 +969,6 @@
 		netif_receive_skb(skb);
 
 next:
-		curr++;
-		curr &= mask;
-
 		done++;
 	}
 
@@ -982,8 +979,6 @@
 	 * Replenish the RX buffer entries.
 	 */
 	while (used < size) {
-		struct ag71xx_buf *buf = &ring->buf[dirty];
-		struct ag71xx_desc *desc = buf->desc;
 		struct sk_buff *skb;
 
 		skb = dev_alloc_skb(rx_buf_size);
@@ -993,15 +988,18 @@
 
 		skb_reserve(skb, rx_buf_offset);
 
-		buf->skb = skb;
-		buf->dma_addr = dma_map_single(&dev->dev, skb->data,
-					       rx_buf_size, DMA_FROM_DEVICE);
+		dirty->skb = skb;
+		dirty->dma_addr = dma_map_single(&dev->dev, skb->data,
+						 rx_buf_size, DMA_FROM_DEVICE);
 
-		desc->data = (u32)buf->dma_addr;
+		desc->data = (u32)dirty->dma_addr;
 		desc->ctrl = DESC_EMPTY;
 
-		dirty++;
-		dirty &= mask;
+		/*
+		 * Move to the next descriptor.
+		 */
+		dirty = dirty->next;
+		desc = dirty->desc;
 
 		used++;
 	}
Index: linux-3.3.8/drivers/net/ethernet/atheros/ag71xx/ag71xx_debugfs.c
===================================================================
--- linux-3.3.8.orig/drivers/net/ethernet/atheros/ag71xx/ag71xx_debugfs.c	2014-06-13 14:29:33.855658779 -0500
+++ linux-3.3.8/drivers/net/ethernet/atheros/ag71xx/ag71xx_debugfs.c	2014-06-13 14:42:18.385049362 -0500
@@ -162,8 +162,6 @@
 	unsigned int len = 0;
 	unsigned long flags;
 	ssize_t ret;
-	int curr;
-	int dirty;
 	u32 desc_hw;
 	int i;
 
@@ -173,30 +171,27 @@
 		return -ENOMEM;
 
 	len += snprintf(buf + len, buflen - len,
-			"Idx ... %-8s %-8s %-8s %-8s . %-10s\n",
-			"desc", "next", "data", "ctrl", "timestamp");
+			"Idx ... %-8s %-8s %-8s %-8s\n",
+			"desc", "next", "data", "ctrl");
 
 	spin_lock_irqsave(&ag->lock, flags);
 
-	curr = (ring->curr % ring->size);
-	dirty = (ring->dirty % ring->size);
 	desc_hw = ag71xx_rr(ag, desc_reg);
 	for (i = 0; i < ring->size; i++) {
 		struct ag71xx_buf *ab = &ring->buf[i];
 		u32 desc_dma = ((u32) ring->descs_dma) + i * ring->desc_size;
 
 		len += snprintf(buf + len, buflen - len,
-			"%3d %c%c%c %08x %08x %08x %08x %c %10lu\n",
+			"%3d %c%c%c %08x %08x %08x %08x %c\n",
 			i,
-			(i == curr) ? 'C' : ' ',
-			(i == dirty) ? 'D' : ' ',
+			(&ring->buf[i] == ring->curr) ? 'C' : ' ',
+			(&ring->buf[i] == ring->dirty) ? 'D' : ' ',
 			(desc_hw == desc_dma) ? 'H' : ' ',
 			desc_dma,
 			ab->desc->next,
 			ab->desc->data,
 			ab->desc->ctrl,
-			(ab->desc->ctrl & DESC_EMPTY) ? 'E' : '*',
-			ab->timestamp);
+			(ab->desc->ctrl & DESC_EMPTY) ? 'E' : '*');
 	}
 
 	spin_unlock_irqrestore(&ag->lock, flags);
